operator/buffer/circular_file.go

```diff
@@ -0,0 +1,327 @@
+package buffer
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"go.uber.org/multierr"
+)
+
+// circularFile is a io.ReadWriteCloser that writes to a fixed length file, such that
+// it wraps around to the beginning when reaching the end.
+// Methods on this struct are not thread-safe and require additional synchronization.
+type circularFile struct {
+	Start      int64
+	ReadPtr    int64
+	End        int64
+	Size       int64
+	Full       bool
+	f          fileLike
+	seekedRead bool
+	seekedEnd  bool
+	closed     bool
+	readPtrEnd bool
+}
+
+var _ io.ReadWriteCloser = (*circularFile)(nil)
+
+func openCircularFile(filePath string, sync bool, size int64) (*circularFile, error) {
+	fileFlags := os.O_CREATE | os.O_RDWR
+	if sync {
+		fileFlags |= os.O_SYNC
+	}
+
+	f, err := os.OpenFile(filePath, fileFlags, 0600)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure that the file, if it already existed, is actually the correct size.
+	fsize, err := f.Seek(0, io.SeekEnd)
+	if err != nil {
+		fCloseErr := f.Close()
+		return nil, multierr.Combine(
```

Comment:
Can we switch to using error wrapping everywhere in this file? Looks like most of stanza does not use `multierr`.

Reply:
I'll have to make some compromise here if we choose not to use multierr. Can't wrap 2 errors, so I can wrap one and string another, or just drop the error (which could be fine in these cases)

We seem to already use this package https://pkg.go.dev/github.com/hashicorp/go-multierror@v1.1.1#section-readme in the key value parser, I could also maybe switch to that?

Reply:
So yeah the key-value parser is the only place I saw we used that. I was also wondering do we care to return to the caller if the file was closed successfully? I know it's correct to capture that error but what do we expect the user to do if we don't close the file? If the program will close on this error the file descriptor will free up anyways.

I'll leave it up to you if we should capture that error but it might help solve wrapping two errors.

Reply:
Yeah, I'll probably just drop the file close error.